### DISCUSSION TITLE
Dissecting pacquet frozen lockfile performance

### DISCUSSION THREAD

**Discussion by KSXGitHub** [pnpm/pnpm#7296] [General]
## Scope

This discussion is centered around the frozen-lockfile feature of pacquet only. It is well known that pacquet algorithm without lockfile is not good enough, but we are not focusing on that.

## Problem

It is easy to see that pacquet algorithm is not parallelized to the fullest, as there are many points where unrelated IO/network operations still happen in serial order. Yet, when I try to parallelize it either by using `rayon` or `tokio::task::spawn`, pacquet becomes slower than before.

## Resources

The branches:
* The `main` branch is at https://github.com/pnpm/pacquet/commit/6b00165bae28c873c6c6e448746c4ece17d9a242 at the time of writing this.
* [`optimization-attempt-rayon`](https://github.com/pnpm/pacquet/tree/optimization-attempt-rayon) is an attempt at using rayon to parallelize some operations. It is at https://github.com/pnpm/pacquet/commit/1f4b10f16a159b15262b92a730a95800762b5470 at the time of writing this.
* [`optimization-attempt-tokio-tasks`](https://github.com/pnpm/pacquet/tree/optimization-attempt-tokio-tasks) is an attempt at using tokio tasks to parallelize some operations. It is at https://github.com/pnpm/pacquet/commit/92d59495d5fd165f10698c61add7cd39b8fe214f at the time of writing this.

The manifest files and lockfiles are located in `tasks/integrated-benchmark/src/fixtures/`.
* Small manifest file and lockfile: Are the default.
* Medium-sized manifest file and lockfile: https://github.com/pnpm/pacquet/commit/8289e36b4ef028648523f7d4a9a2efbb13ad1ed7

To compare benchmark against the main branch, run:

```bash
just integrated-benchmark -s frozen-lockfile $TARGET_BRANCH main -V
````

**NOTE:** The benchmarks are using a small lockfile by default. You should manually copy the medium-size lockfile and manifest file for a more objective measurement. Remember to manually set `autoInstallPeers` in the lockfile to `false`.

**NOTE:** You can add `--with-pnpm` to compare pacquet against pnpm. Be aware that startup time will hold back pnpm if using the default lockfile, so a bigger lockfile is necessary to be objective.

**Comment by kdy1** [comment]
I want to attach the profiler to the command. What are `just` and `just integrated-benchmark`?

**Reply by KSXGitHub** [reply]
[just](https://github.com/casey/just) is a task runner. All the tasks are stored in `justfile` in the root of the repo.

It runs a Rust program at `tasks/integrated-benchmark`.

**Comment by kdy1** [comment]
I think we need a benchmark that is more reliable.

**Reply by KSXGitHub** [reply]
Is the lockfile not big enough, or something else?

**Reply by kdy1** [reply]
It varies too much. It seems like it has some network operations.

**Reply by KSXGitHub** [reply]
> It varies too much. It seems like it has some network operations.

All network operations only connect to localhost which is served by [verdaccio](https://github.com/verdaccio/verdaccio).

**Reply by kdy1** [reply]
But it still varies from 150ms to 650ms.

**Reply by KSXGitHub** [reply]
My machine (Linux, 56 cores, 48GB of RAM) is quite stable.

<details><summary>bench-work-env/BENCHMARK_REPORT.json</summary>

```json
{
  "results": [
    {
      "command": "pacquet@HEAD",
      "mean": 0.13035584404,
      "stddev": 0.008067993213068583,
      "median": 0.12837198564,
      "user": 0.6484349066666666,
      "system": 0.5851663733333333,
      "min": 0.11587404964,
      "max": 0.14220072864,
      "times": [
        0.13805841264000002,
        0.13784145564,
        0.14220072864,
        0.13894517864,
        0.12523665964,
        0.13909586864,
        0.12507481564,
        0.12837198564,
        0.12144884164,
        0.12672492064000002,
        0.12493211264000001,
        0.11587404964,
        0.12384233564,
        0.13879491864000001,
        0.12889537664
      ],
      "exit_codes": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0
      ]
    },
    {
      "command": "pacquet@main",
      "mean": 0.12817658916941177,
      "stddev": 0.007999823319031376,
      "median": 0.12709307764,
      "user": 0.4304567694117647,
      "system": 0.46009327529411764,
      "min": 0.11347365664,
      "max": 0.13927691164,
      "times": [
        0.12102636863999999,
        0.12533512664000002,
        0.11468248464,
        0.12001481564000001,
        0.12709307764,
        0.11347365664,
        0.12586417164,
        0.12999639964,
        0.12615296564,
        0.13927691164,
        0.13159887164,
        0.13921117664000002,
        0.13567371164,
        0.13893893164,
        0.13319960764,
        0.12477131363999999,
        0.13269242464
      ],
      "exit_codes": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0
      ]
    }
  ]
}
```

</details>

Anyway, the result observed in your machine could indicate a problem with the implementation itself. Do you think mixing tokio and rayon would create unpredictable results?

**Reply by kdy1** [reply]
I think the difference here is mac vs linux problem.

Many project mixes rayon and tokio, and itself is fine. But we need to restrict the scope of operations of each threadpool to make it performant. The characteristics of tasks that each thread pool should be similar (blocking without cpu / cpu-heavy / network io) and the pool size should be dtermined based on it.

**Reply by KSXGitHub** [reply]
Does this mean that we have to balance between tokio thread pool and rayon thread pool? But then how could one determine the right balance? Maybe we should migrate the codebase to only use tokio or rayon, not both?

**Comment by zkochan** [comment]
This has been solved. Pacquet is now faster than pnpm.
